### DISCUSSION TITLE
Python could not be found on Windows

### DISCUSSION THREAD

**Discussion by Dioskurides** [vladmandic/sdnext#373] [Announcements]
### Issue Description

Hi, i cant launch webui.bat im getting the following error: 

Couldn't launch python

exit code: 9009

stderr:
The command "C:\Users\rouve\AppData\Local\Programs\Python\Python311" is either misspelled or
could not be found.

Launch unsuccessful. Exiting.
Press any key . . .

What can i do to fix this? 

Thanks in advance

### Version Platform Description

Im using win 11 and python 3.11.3

**Comment by kaelysdev** [comment]
Python 3.11 is not supported, use 3.10 instead

**Comment by Dioskurides** [comment]
> Python 3.11 is not supported, use 3.10 instead

I tried, it didnt work either

**Comment by ShyRed** [comment]
> > Python 3.11 is not supported, use 3.10 instead
> 
> I tried, it didnt work either

What error message are you getting for Python 3.10?

What happens if you open a command prompt (type `cmd` in windows start menu) and inside the command prompt you type `python`?

**Reply by Dioskurides** [reply]
> > > Python 3.11 is not supported, use 3.10 instead
> > 
> > 
> > I tried, it didnt work either
> 
> What error message are you getting for Python 3.10?
> 
> What happens if you open a command prompt (type `cmd` in windows start menu) and inside the command prompt you type `python`?

The error is exactly the same with python 3.10 and if i type python it says 

C:\Users\rouve>py
Python 3.10.10 (tags/v3.10.10:aad5f6a, Feb  7 2023, 17:20:36) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.

**Reply by ShyRed** [reply]
Seems that your PATH variable is set correctly, otherwise it wouldn't have been able to load up python in the command prompt.

You said that the error is exactly the same with python 3.10? Does that mean that it is still writing `The command "C:\Users\rouve\AppData\Local\Programs\Python\Python311" is either misspelled or
could not be found.` in the error message? Still looking for python 3.11?

If that is the case I can imagine that something went wrong when it created your venv using Python 11.
It might be worth a try to rename your `venv` folder inside the `automatic` folder to something like `venv_backup` and have it re-create the venv on next webui start.

**Reply by Dioskurides** [reply]
> Seems that your PATH variable is set correctly, otherwise it wouldn't have been able to load up python in the command prompt.
> 
> You said that the error is exactly the same with python 3.10? Does that mean that it is still writing `The command "C:\Users\rouve\AppData\Local\Programs\Python\Python311" is either misspelled or could not be found.` in the error message? Still looking for python 3.11?
> 
> If that is the case I can imagine that something went wrong when it created your venv using Python 11. It might be worth a try to rename your `venv` folder inside the `automatic` folder to something like `venv_backup` and have it re-create the venv on next webui start.

Yeah its still this one, the thing is i had 3.10 before i installed 3.11 and only installed 3.11 because i saw the the path ended with python311 i figured it would be worth a try installing 3.11 idk if thats important but i translated the error im running german windows. Thats the original error now back under 3.10 again:

Couldn't launch python

exit code: 9009

stderr:
Der Befehl "C:\Users\rouve\AppData\Local\Programs\Python\Python311" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.

Launch unsuccessful. Exiting.
Drücken Sie eine beliebige Taste . . .

**Reply by Dioskurides** [reply]
> Seems that your PATH variable is set correctly, otherwise it wouldn't have been able to load up python in the command prompt.
> 
> You said that the error is exactly the same with python 3.10? Does that mean that it is still writing `The command "C:\Users\rouve\AppData\Local\Programs\Python\Python311" is either misspelled or could not be found.` in the error message? Still looking for python 3.11?
> 
> If that is the case I can imagine that something went wrong when it created your venv using Python 11. It might be worth a try to rename your `venv` folder inside the `automatic` folder to something like `venv_backup` and have it re-create the venv on next webui start.

And i dont have the venv folder i just git pulled and tried installing which didnt work and just resulted in the error

**Reply by vladmandic** [reply]
if you run `git clone` followed by `webui.bat`, it would automatically create `venv`

**Reply by Dioskurides** [reply]
> if you run `git clone` followed by `webui.bat`, it would automatically create `venv`

Yeah except it didnt and simply threw the error im talking about.

**Reply by ShyRed** [reply]
Can you type `%PYTHON%` in command prompt and post the result? The batch file checks for an environment variable called PYTHON it seems. Maybe PYTHON points to the wrong `C:\Users\rouve\AppData\Local\Programs\Python\Python311` while PATH points to the correct one.

Solution would be modifying/deleting your PYTHON environment variable in your OS settings.

**Reply by Dioskurides** [reply]
> Can you type `%PYTHON%` in command prompt and post the result? The batch file checks for an environment variable called PYTHON it seems. Maybe PYTHON points to the wrong `C:\Users\rouve\AppData\Local\Programs\Python\Python311` while PATH points to the correct one.
> 
> Solution would be modifying/deleting your PYTHON environment variable in your OS settings.

Sure, i get the same error as before except it dosent have the error code listed above: C:\Users\rouve>%PYTHON%
Der Befehl "C:\Users\rouve\AppData\Local\Programs\Python\Python311" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.

**Reply by ShyRed** [reply]
The issue is your `PYTHON` environment variable, then. You can delete the variable in your windows system settings.

**Reply by Dioskurides** [reply]
> The issue is your `PYTHON` environment variable, then. You can delete the variable in your windows system settings.

Im sorry but could tell me how i do that? :D

**Reply by dilectiogames** [reply]
you should try watching some youtube videos tutorials about changing updating paths BUT careful with those options, if you mess something there you can break windows

**Reply by ShyRed** [reply]
1. On Windows desktop, click Start > Control Panel > System. ...
2. Click the Advanced tab, and click Environment Variables.
3. Under System Variables, select the PYTHON variable, and click Delete.
4. Click OK.

Just to avoid confusion: I'm talking about the `PYTHON` variable. Do not touch the `PATH` variable.

**Reply by Dioskurides** [reply]
> you should try watching some youtube videos tutorials about changing updating paths BUT careful with those options, if you mess something there you can break windows

This is why i wont watch a static vid but prefer to ask someone who knows what they are talking about :) Just like you did here

**Reply by ShyRed** [reply]
![image](https://user-images.githubusercontent.com/1138805/233791590-0febe473-8926-4c94-a92c-ad4902265479.png)

**Reply by Dioskurides** [reply]
Its the same er

> 1. On Windows desktop, click Start > Control Panel > System. ...
> 2. Click the Advanced tab, and click Environment Variables.
> 3. Under System Variables, select the PYTHON variable, and click Delete.
> 4. Click OK.
> 
> Just to avoid confusion: I'm talking about the `PYTHON` variable. Do not touch the `PATH` variable.

Thank you very much, i just did that, but i still get the error unfortunatly.

**Reply by ShyRed** [reply]
Can you type `%PYTHON%` in command prompt again?

It should tell you
> '%PYTHON%' is not recognized as an internal or external command,
operable program or batch file.

**Reply by Dioskurides** [reply]
> ![image](https://user-images.githubusercontent.com/1138805/233791590-0febe473-8926-4c94-a92c-ad4902265479.png)

<img width="437" alt="2023-04-22" src="https://user-images.githubusercontent.com/128710762/233791748-d8c2345e-40c1-4961-9a0d-6f04257ba47a.png">

**Reply by ShyRed** [reply]
Did you check "Benutzervariablen for rouve" and "Systemvariablen"? PYTHON should not exist in both of them.

**Reply by Dioskurides** [reply]
> Did you check "Benutzervariablen for rouve" and "Systemvariablen"? PYTHON should not exist in both of them.

Yup, its gone in both

**Reply by ShyRed** [reply]
And command prompt is still pretending PYTHON exists? Can you restart your computer and try again with webui.bat?

**Reply by Dioskurides** [reply]
> And command prompt is still pretending PYTHON exists? Can you restart your computer and try again with webui.bat?

ofc im glad someone is helping me ill tell you asap :)

**Reply by Dioskurides** [reply]
> And command prompt is still pretending PYTHON exists? Can you restart your computer and try again with webui.bat?

I restarted and ran %PYTHON% in cmd still the same error

**Reply by Dioskurides** [reply]
> And command prompt is still pretending PYTHON exists? Can you restart your computer and try again with webui.bat?

I just checked the variables again (after i ran %PYTHON) and noticed that the PYTHON environment variable reappeared in my settings

**Reply by ShyRed** [reply]
[mywebui.txt](https://github.com/vladmandic/automatic/files/11301624/mywebui.txt)

You can try and place that file next  to your webui.bat and rename the mywebui.txt to mywebui.bat.

It should override the PYTHON environment variable. It's just a workaround and no fix, but at least you should be able to use automatic for the time being.

**Reply by Dioskurides** [reply]
> And command prompt is still pretending PYTHON exists? Can you restart your computer and try again with webui.bat?

Update: i just removed it again and now webui.bat complains that i dont have torch which is odd since im already using automatic1111 
<img width="854" alt="2023-04-22 (1)" src="https://user-images.githubusercontent.com/128710762/233792946-7ee75430-0f07-4968-809a-c0c41033eae1.png">

**Reply by ShyRed** [reply]
Try starting webui.bat with `--upgrade` parameter by typing `webui.bat --upgrade` in your command prompt inside your automatic folder.

If nothing helps, delete your `venv` folder to force a complete re-install. I do not know how much Python 11 messed with the venv and if it can be recovered.

**Reply by Dioskurides** [reply]
> Try starting webui.bat with `--upgrade` parameter by typing `webui.bat --upgrade` in your command prompt inside your automatic folder.
> 
> If nothing helps, delete your `venv` folder to force a complete re-install. I do not know how much Python 11 messed with the venv and if it can be recovered.

This seems to be working its installing the packages now thank you, you are a genius!

**Comment by vladmandic** [comment]
this is an issue with python installation not with repo, moving to discussions. hopefully someone can help there.

**Comment by dilectiogames** [comment]
I've got issues with python not being found because I've had various versions installed. for some unknown reason even if python 3 had the path setup in the system variables, it returned a link to the older version instead of 3.1. or 3.11

solution is to uninstall the older version, that fixed finding python. but now remains a different issue, 

I get all these issues when starting webui.bat

here is a part of them

4:57:35-327407 INFO     Installing package: facexlib
14:57:49-599602 ERROR    Error running pip with args: install --upgrade facexlib
14:57:49-604602 INFO     Installing package: gfpgan
14:57:57-725473 ERROR    Error running pip with args: install --upgrade gfpgan
14:57:57-736053 INFO     Installing package: realesrgan
14:58:06-409932 ERROR    Error running pip with args: install --upgrade realesrgan
14:58:06-422102 INFO     Installing packages
14:58:06-428649 INFO     Installing repositories
14:58:06-739382 INFO     Installing submodules
14:58:10-589120 ERROR    Error running git with args: submodule --quiet update --init --recursive
14:58:10-594086 INFO     Updating submodules
14:58:13-364574 ERROR    Error running git with args: checkout main
14:58:13-426285 ERROR    Error running git with args: pull --rebase --autostash
1

**Reply by vladmandic** [reply]
the fact that it complains about those 3 specific packages tells me you're still using python 3.11 that is unsupported.
check `setup.log` for more details.

**Reply by dilectiogames** [reply]
is 3.11, just trusting youtubers that don't know what they are doing. I also have the 3.10 but how can I make it use that version?

the older version of python was 2.7. when that was installed it complained there were o python at all

**Reply by vladmandic** [reply]
> I also have the 3.10 but how can I make it use that version?

Whichever is first found in path. There are also utilities that allow to switch multiple side-by-side installs of Python, but I don't use them, so wouldn't know.

**Reply by dilectiogames** [reply]
I will uninstall all of them and add only the version 10. But I've forgot to add that even with those errors the webui did started a couple of times and generated images. It took 6 minutes to start but did start

**Comment by dilectiogames** [comment]
I've got rid of all pythons installations, and installed only the python 10, but now webui bat ask for python 11

venv "...\Vlad1111\automatic\venv\Scripts\Python.exe"
No Python at '"...\AppData\Local\Programs\Python\Python311\python.exe'
Press any key to continue . . .

in windows cmd writing python gives correct version

E:\>python
Python 3.10.11 (tags/v3.10.11....

I've checked the path values and I have only the pyhton 10 paths there

**Reply by ShyRed** [reply]
Can you check if you have a PYTHON environment variable set that is still pointing to the old Python 11 install?

**Reply by dilectiogames** [reply]
I've did that, it points to 3.10. in windows console python gives correct versions and starts correctly.

I guess inside the installation files of this app somewhere it was written down the path for 3.11 and keeps using that instead of windows path

**Comment by dilectiogames** [comment]
I've fixed the issue with asking for python 3.11 by updating the info found in the file pyvenv.cfg which is located inside venv folder

all the data in that file pointed to 3.11.

**Reply by rugabunda** [reply]
If that works, great, but the venv folder still was created by python 311, using my method below, you create a 310 venv, for whatever its worth

https://github.com/vladmandic/automatic/discussions/373#discussioncomment-6758114

**Comment by dilectiogames** [comment]
but now I get thise error :( ahhhhhhhhhh!



16:47:18-518613 INFO     Python 3.10.11 on Windows
16:47:18-800757 INFO     Running setup
16:47:18-871629 INFO     Version: e6e296a3 Sat Apr 22 09:05:26 2023 -0400
16:47:23-970343 INFO     Latest published version: cb465b12ddc5e4b5f6566030021a26630b927ba6 2023-04-22T13:37:17Z
16:47:23-973730 INFO     Updating Wiki
16:47:28-873371 INFO     nVidia toolkit detected
16:47:38-303070 INFO     Installing requirements
16:47:38-324020 INFO     Installing package: facexlib
16:48:14-648843 INFO     Installing package: gfpgan
16:48:25-269089 INFO     Installing package: realesrgan
16:48:28-183805 INFO     Installing packages
16:48:28-189805 INFO     Installing repositories
16:48:28-698183 INFO     Installing submodules
16:48:33-820180 ERROR    Error running git with args: submodule --quiet update --init --recursive
16:48:33-823810 INFO     Updating submodules
16:48:36-669820 ERROR    Error running git with args: checkout main
16:48:36-739341 ERROR    Error running git with args: pull --rebase --autostash
16:49:00-207220 INFO     Extensions enabled: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'prompt-bracket-checker', 'ScuNET',
                         'sd-dynamic-thresholding', 'sd-extension-aesthetic-scorer', 'sd-extension-steps-animation',
                         'sd-extension-system-info', 'sd-webui-controlnet', 'sd-webui-model-converter', 'seed_travel',
                         'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
16:49:00-338187 ERROR    Error running git with args: checkout main
16:49:00-403251 ERROR    Error running git with args: pull --rebase --autostash
16:50:29-759493 INFO     Extensions enabled: []
16:50:46-250061 WARNING  Setup complete with errors (5)
16:50:46-253954 INFO     Setting environment tuning
16:50:46-255538 INFO     Server arguments: []
┌───────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────┐
│ D:\AI\Vlad1111\automatic\launch.py:99 in <module>                                                                   │
│                                                                                                                     │
│    98     logging.disable(logging.INFO)                                                                             │
│ >  99     import webui                                                                                              │
│   100     webui.webui()                                                                                             │
│                                                                                                                     │
│ D:\AI\Vlad1111\automatic\webui.py:12 in <module>                                                                    │
│                                                                                                                     │
│    11                                                                                                               │
│ >  12 import torch # pylint: disable=C0411                                                                          │
│    13 import torchvision # pylint: disable=W0611,C0411                                                              │
│                                                                                                                     │
│ D:\AI\Vlad1111\automatic\venv\lib\site-packages\torch\__init__.py:133 in <module>                                   │
│                                                                                                                     │
│    132                 err.strerror += f' Error loading "{dll}" or one of its dependencies.'                        │
│ >  133                 raise err                                                                                    │
│    134                                                                                                              │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
OSError: [WinError 126] The specified module could not be found. Error loading
"D:\AI\Vlad1111\automatic\venv\lib\site-packages\torch\lib\torch_python.dll" or one of its dependencies.
Press any key to continue . . .

**Comment by teebu** [comment]
Fresh install, same error: 

```
stderr:
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
```

My solution was edit the webui.bat and add near the top, point to your python installation path.

`set PYTHON="C:\Program Files\Python310\python.exe"`

**Comment by s-a** [comment]
hacked `webui.bat` with 

```shell
rem if not defined PYTHON 
(set PYTHON="c:\Users\WINDOWS_USERNAME\MiniConda3\envs\vlad\python.exe")
```

**Comment by nemilya** [comment]
Was same problem `3.11` python was initialized in `venv`, fast solution - change in `webui.sh`:

```diff
diff --git a/webui.sh b/webui.sh
index ebbe586f..8289f277 100755
--- a/webui.sh
+++ b/webui.sh
@@ -35,7 +35,7 @@ fi
 # python3 executable
 if [[ -z "${python_cmd}" ]]
 then
-    python_cmd="python3"
+    python_cmd="python3.10"
 fi

 # git executable
```

**Reply by vladmandic** [reply]
you don't need to hack it, just define `python_cmd` environment variable before runing `webui`.

**Comment by axelaxolotl** [comment]
I fixed it for me. windows installed the python 3.11 app, check in the windows store if it is installed. I did not see it under installed programms. after that go to enviromental variables and move the microsoft store entry in your users path variable to the bottom and add one linking to your python 3.10.x over it. only found this out because without installed python, using the python command 3.11 would launch. Reinstalling python normally will not help

this is the problem but changing this alone will not help you have to do as I instructed + uninstall python 3.11 app if you see it in the store and what is done in the link bellow 
https://stackoverflow.com/questions/58754860/cmd-opens-windows-store-when-i-type-python

Edit: the python to store link only happens after you have uninstalled the app python 3.11

(this post is in bad english but better than nothing)

**Comment by rugabunda** [comment]
Delete your venv folder in \automatic\,

head to your 3.10 python folder at command line, `C:\users\username\AppData\Local\Programs\Python\Python310` 

Run this in powershell/md:

` .\python.exe -m venv D:\#AI\automatic\venv `

(replace `D:\#AI\automatic\venv` with the path of what will become your new  automatic\venv folder
